File: typescript/src/n0235_lowest_common_ancestor_of_a_binary_search_tree.ts
import type { TreeNode } from './utils/tree';

export function lowestCommonAncestor(
  root: TreeNode | null,
  p: TreeNode | null,
  q: TreeNode | null
): TreeNode | null {
  const min = Math.min(p!.value, q!.value);
  const max = Math.max(p!.value, q!.value);
  let node = root;
  while (node) {
    if (max < node.value) {
      node = node.left;
    } else if (min > node.value) {
      node = node.right;
    } else {
      break;
    }
  }
  return node;
}

File: rust/src/n0235_lowest_common_ancestor_of_a_binary_search_tree.rs
use std::{cell::RefCell, rc::Rc};

use crate::utils::tree::TreeNode;

pub fn lowest_common_ancestor(
    root: Option<Rc<RefCell<TreeNode>>>,
    p: Option<Rc<RefCell<TreeNode>>>,
    q: Option<Rc<RefCell<TreeNode>>>,
) -> Option<Rc<RefCell<TreeNode>>> {
    let p_value = p.as_ref()?.borrow().value;
    let q_value = q.as_ref()?.borrow().value;
    let min = p_value.min(q_value);
    let max = p_value.max(q_value);
    let mut cursor = root?;
    loop {
        let cursor_value = cursor.borrow().value;
        cursor = if max < cursor_value {
            cursor.borrow().left.clone()?
        } else if min > cursor_value {
            cursor.borrow().right.clone()?
        } else {
            return Some(cursor);
        }
    }
}
